pkg/sql/opt/norm/testdata/rules/with

```diff
@@ -125,237 +125,74 @@ values
 
-norm expect=InlineWith
+# We should inline both foo and bar.
+norm expect=(InlineWith,InlineWithScan)
 WITH foo AS (SELECT 1), bar AS (SELECT 2) SELECT (SELECT * FROM foo) + (SELECT * FROM bar) + (SELECT * FROM bar)
 ----
-with &2 (bar)
+values
  ├── columns: "?column?":6
  ├── cardinality: [1 - 1]
  ├── immutable
  ├── key: ()
  ├── fd: ()-->(6)
- ├── values
- │    ├── columns: "?column?":2!null
- │    ├── cardinality: [1 - 1]
- │    ├── key: ()
- │    ├── fd: ()-->(2)
- │    └── (2,)
- └── values
-      ├── columns: "?column?":6
-      ├── cardinality: [1 - 1]
-      ├── immutable
-      ├── key: ()
-      ├── fd: ()-->(6)
-      └── tuple
-           └── plus
-                ├── plus
-                │    ├── subquery
-                │    │    └── values
-                │    │         ├── columns: "?column?":3!null
-                │    │         ├── cardinality: [1 - 1]
-                │    │         ├── key: ()
-                │    │         ├── fd: ()-->(3)
-                │    │         └── (1,)
-                │    └── subquery
-                │         └── with-scan &2 (bar)
-                │              ├── columns: "?column?":4!null
-                │              ├── mapping:
-                │              │    └──  "?column?":2 => "?column?":4
-                │              ├── cardinality: [1 - 1]
-                │              ├── key: ()
-                │              └── fd: ()-->(4)
-                └── subquery
-                     └── with-scan &2 (bar)
-                          ├── columns: "?column?":5!null
-                          ├── mapping:
-                          │    └──  "?column?":2 => "?column?":5
-                          ├── cardinality: [1 - 1]
-                          ├── key: ()
-                          └── fd: ()-->(5)
+ └── tuple
+      └── plus
+           ├── plus
+           │    ├── subquery
+           │    │    └── values
+           │    │         ├── columns: "?column?":3!null
+           │    │         ├── cardinality: [1 - 1]
+           │    │         ├── key: ()
+           │    │         ├── fd: ()-->(3)
+           │    │         └── (1,)
+           │    └── subquery
+           │         └── values
+           │              ├── columns: "?column?":4!null
+           │              ├── cardinality: [1 - 1]
+           │              ├── key: ()
+           │              ├── fd: ()-->(4)
+           │              └── (2,)
+           └── subquery
+                └── values
+                     ├── columns: "?column?":5!null
+                     ├── cardinality: [1 - 1]
+                     ├── key: ()
+                     ├── fd: ()-->(5)
+                     └── (2,)
 
-# We should inline foo, but not bar.
-norm expect=InlineWith
+# We should inline both foo and bar.
+norm expect=(InlineWith,InlineWithScan)
 WITH foo AS (SELECT 1), bar AS (SELECT 2) SELECT * FROM foo CROSS JOIN bar CROSS JOIN bar AS bar2
 ----
-with &2 (bar)
+values
```

Comment:
In this example, inlining `bar` is clearly more efficient than not inlining `bar`. But if `bar` was a costly relational expression, inlining could be less efficient because we'd be duplicating that costly work. For that reason, I'm not convinced we can inline a WITH expression referenced more than once in a normalization rule. We could restrict `InlineWithScan` to only apply to a single reference for now.